storage/src/main/scala/coop/rchain/storage/StorageActions.scala

```diff
@@ -0,0 +1,118 @@
+package coop.rchain.storage
+
+import cats.implicits._
+import coop.rchain.storage.util.ignore
+
+import scala.annotation.tailrec
+
+// TODO(ht): improve docstrings
+trait StorageActions {
+
+  /* Consume */
+
+  @tailrec
+  private[storage] final def findMatchingDataCandidates[C, P, A](
+      data: List[(A, Int)],
+      pattern: P,
+      channel: C)(implicit m: Match[P, A]): Option[(A, C, Int)] =
+    data match {
+      case Nil => None
+      case (a, i) :: as =>
+        m.get(pattern, a) match {
+          case None      => findMatchingDataCandidates(as, pattern, channel)
+          case Some(mat) => Some((mat, channel, i))
```

Comment:
Just a comment that we are returning the first available match here. We could return all possible matches and then pick the "best" one with a performance hit.

Reply:
@guardbotmk3: thoughts?

Reply:
Btw to be clear my position is against doing this as of now. In the future this might be a more enticing option.

Reply:
@KentShikama: could you be a little more specific about which "this" you are against?

Reply:
I'm against returning all possible matching data candidates and then filtering to pick the "best" one. In another words, I like the code as you wrote it. I just wanted to throw this out there in case in the future something changes about Rchain and someone looks at this PR.

Reply:
Let's let the non-determinism come from write ordering and threads. I think this should just return first match.